rust/src/libdnf_sys.rs

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use libc;
+use glib_sys;
+
+/* This is an experiment in calling libdnf functions from Rust. Really, it'd be more sustainable to
+ * generate a "libdnf-sys" from SWIG (though it doesn't look like that's supported yet
+ * https://github.com/swig/swig/issues/1468) or at least `bindgen` for the strict C parts of the
+ * API. For now, I just took the shortcut of manually defining a tiny subset we care about. */
+
+#[repr(C)]
+pub(crate) struct DnfPackage(*mut libc::c_void);
```

Comment:
I trust you this is correct, I'm having an hard time finding the real type definition of this in libdnf.

Reply:
It's an opaque structure. I'm mostly following what glib does here, though I may have been looking at an older example. The latest one uses just `libc::c_void`. See e.g. `GHashTable` in https://docs.rs/glib-sys/0.9.0/src/glib_sys/lib.rs.html#1266.

The nomicon though suggests using a struct with a private field so it can't ever be instantiated on the Rust side: https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs. Will switch to that.

Reply:
Oh actually, `DnfPackage(libc::c_void)` accomplishes this as well. Switched to that in latest! :arrow_down: